File: android/feature/catalog/src/test/kotlin/SeriesResumeTest.kt
package catalog

import model.Kind
import model.MediaSet
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNull

/** Mirrors `web/test/series-resume.test.ts`, case for case. */
class SeriesResumeTest {
    private fun ep(
        season: Int,
        episode: Int,
    ): MediaSet =
        MediaSet(
            setId = "s${season}e$episode", kind = Kind.EPISODE, title = "Ep $season.$episode", show = "Show",
            chapter = null, path = null, season = season, episodeFirst = episode, episodeLast = null,
            year = null, durationSecs = null, posterPath = null, totalBytes = 0,
        )

    private val show =
        listOf(
            Division("Season 1", season = 1, items = listOf(ep(1, 1), ep(1, 2)), children = emptyList()),
            Division("Season 2", season = 2, items = listOf(ep(2, 1), ep(2, 2)), children = emptyList()),
        )

    private val none: (String) -> Double? = { null }
    private val nothingWatched: (String) -> Boolean = { false }

    @Test
    fun aFreshShowStartsAtItsFirstEpisode() {
        val pick = seriesResume(show, none, emptyList(), nothingWatched)!!
        assertEquals("s1e1", pick.set.setId)
        assertEquals(ResumeVerb.PLAY, pick.verb)
        assertNull(pick.at)
    }

    @Test
    fun theEpisodeStoppedPartwayThroughWinsEvenBehindAFinishedOne() {
        val pick =
            seriesResume(
                show,
                resumeOf = { if (it == "s1e2") 600.0 else null },
                recent = listOf("other-show-ep", "s1e2"),
                watched = { it == "s2e1" },
            )!!
        assertEquals("s1e2", pick.set.setId)
        assertEquals(ResumeVerb.RESUME, pick.verb)
        assertEquals(600.0, pick.at)
    }

    @Test
    fun otherwiseTheOneAfterTheFurthestFinishedAcrossASeason() {
        val pick = seriesResume(show, none, emptyList()) { it == "s1e1" || it == "s1e2" }!!
        assertEquals("s2e1" to ResumeVerb.CONTINUE, pick.set.setId to pick.verb)
    }

    @Test
    fun aShowWatchedToTheEndOffersItsStartAgain() {
        val pick = seriesResume(show, none, emptyList()) { true }!!
        assertEquals("s1e1" to ResumeVerb.PLAY, pick.set.setId to pick.verb)
    }

    @Test
    fun anEmptyShowOffersNothing() {
        assertNull(seriesResume(emptyList(), none, emptyList(), nothingWatched))
    }

    @Test
    fun theShortLabel() {
        assertEquals("S3 E15", episodeShort(ep(3, 15)))
        assertEquals(
            "Pilot",
            episodeShort(
                MediaSet(
                    setId = "p", kind = Kind.EPISODE, title = "Pilot", show = null, chapter = null, path = null,
                    season = null, episodeFirst = null, episodeLast = null, year = null, durationSecs = null,
                    posterPath = null, totalBytes = 0,
                ),
            ),
        )
    }
}

File: web/test/series-resume.test.ts
import { describe, expect, test } from "bun:test";
import { episodeShort, seriesResume } from "../public/lib/catalog/series-resume.js";

const ep = (season: number, episode: number) => ({ setId: `s${season}e${episode}`, kind: "ep", show: "Show", season, episode: String(episode), title: `Ep ${season}.${episode}` });
// Only the fields the rule reads; a real set carries many more.
const show = {
  divisions: [
    { title: "Season 1", season: 1, items: [ep(1, 1), ep(1, 2)], children: [] },
    { title: "Season 2", season: 2, items: [ep(2, 1), ep(2, 2)], children: [] },
  ],
} as unknown as Parameters<typeof seriesResume>[0];
const none = { resumeOf: () => null, recent: [] as string[], watched: () => false };

describe("where a show carries on", () => {
  test("a fresh show starts at its first episode", () => {
    const pick = seriesResume(show, none)!;
    expect([pick.set.setId, pick.verb, pick.at]).toEqual(["s1e1", "Play", null]);
  });

  test("the episode stopped partway through wins, even behind a finished one", () => {
    const pick = seriesResume(show, {
      resumeOf: (id) => (id === "s1e2" ? 600 : null),
      recent: ["other-show-ep", "s1e2"],
      watched: (id) => id === "s2e1",
    })!;
    expect([pick.set.setId, pick.verb, pick.at]).toEqual(["s1e2", "Resume", 600]);
  });

  test("otherwise the one after the furthest finished, across a season", () => {
    const pick = seriesResume(show, { ...none, watched: (id) => id === "s1e1" || id === "s1e2" })!;
    expect([pick.set.setId, pick.verb]).toEqual(["s2e1", "Continue"]);
  });

  test("a show watched to the end offers its start again", () => {
    const pick = seriesResume(show, { ...none, watched: () => true })!;
    expect([pick.set.setId, pick.verb]).toEqual(["s1e1", "Play"]);
  });

  test("an empty show offers nothing", () => {
    expect(seriesResume({ divisions: [] } as unknown as Parameters<typeof seriesResume>[0], none)).toBeNull();
  });

  test("the short label", () => {
    expect(episodeShort(ep(3, 15))).toBe("S3 E15");
    expect(episodeShort({ title: "Pilot" } as never)).toBe("Pilot");
  });
});
